lib/main/atom/views/outline/navigationTreeComponent.tsx

```diff
@@ -0,0 +1,409 @@
+    const modifiersClasses = this.getClassForKindModifiers(navTree.kindModifiers)
+    if (modifiersClasses) {
+      navTree.styleClasses += " " + modifiersClasses
+    }
+
+    if (navTree.childItems) {
+      if (navTree.childItems.length < 1) {
+        // normalize: remove empty lists
+        navTree.childItems = undefined
+        return
+      }
+
+      // TODO should there be a different sort-order?
+      //     for now: sort ascending by line-number
+      navTree.childItems.sort((a, b) => this.getNodeStartLine(a) - this.getNodeStartLine(b))
+
+      let child: NavigationTreeViewModel
+      for (child of navTree.childItems) {
+        this.prepareNavTree(child)
+      }
+    }
+  }
+
+  /**
+   * HELPER transfere collapsed state from old NavigationTreeViewModel to
+   * new view model.
+   *
+   * @returns {boolean} TRUE, if newTree and oldTree matched title
+   */
+  private restoreCollapsed(
+    newTree: NavigationTreeViewModel | null,
+    oldTree: NavigationTreeViewModel | null,
+  ): boolean {
+    if (!newTree || !oldTree) return newTree === oldTree
+
+    // a bit of guess work here:
+    // there may have been additions/deletions in the children
+    // (in comparision to the previous navTree), so the tranfere of
+    // the collapsed state really is a heuristical process.
+    //
+    // For now, we assume, if the name (i.e. node.text) is the same
+    // then it refers to the same entity (i.e. it should get the same
+    // collapsed state); which is not true in case a variable/function/etc
+    // was renamed.
+    // But we do not want the get too elaborate and do expensive modification-
+    // detection here, so in case of renaming, we just reset the collapsed
+    // state to the default (i.e. expanded).
+    // Same for reordering etc. of children: for complex changes we just
+    // revert to the default state.
+
+    if (newTree.text === oldTree.text) {
+      if (oldTree.collapsed) {
+        newTree.collapsed = true
+      }
+
+      if (newTree.childItems && oldTree.childItems) {
+        let newChild: NavigationTreeViewModel
+        let oldChild: NavigationTreeViewModel
+        for (let i = 0, size = newTree.childItems.length; i < size; ++i) {
+          newChild = newTree.childItems[i]
+          oldChild = oldTree.childItems[i]
+          // allow for one addition / deletion in the children
+          // (i.e. check if there's a match in the previous/next position)
+          if (!this.restoreCollapsed(newChild, oldChild)) {
+            // try, if a child was removed
+            oldChild = oldTree.childItems[i + 1]
+            if (!this.restoreCollapsed(newChild, oldChild)) {
+              // try, if a child was added
+              oldChild = oldTree.childItems[i - 1]
+              this.restoreCollapsed(newChild, oldChild)
+            }
+          }
+        }
+      }
+      return true
+    }
+
+    return false
+  }
+
+  private whileRendering: {lastCursorLine: number | null} = {
```

Comment:
well, it's a leftover from the previous implementation that I used as a starting point

I am not 100% sure I understand it completely, but I think it's purpose is to guard against the case that (a) currently there is no editor set and (b) there is no cursor set yet for the editor

I never tested, if (a) or (b) could really happen (since I just kept using the old code fragment)

Reply:
(b) isn't really possible, at least from my experience. (a) is not only possible, but it actually happens, since `render()` is called before the first call to `subscribeToEditor`. But I believe there has to be a more elegant way of handling it than this.

Reply:
I modified this a bit:  
extended the object to `PositionState` that hold the cursor-position and the `selectedNode` and passing it down the `NavigationNodeComponent`

-> so now, in `NavigationNodeComponent` the reference to the partent/root `NavigationTreeComponent` is only needed for `gotoNode(node)` upon a `click` event.

I am not sure, how to best get rid of this:  
I thought about an EventEmitter, but then you would need to link the EventEmitters of the `NavigationNodeComponent` so that they pass the event to their parent until root `NavigationTreeComponent` is reached.  
That doesn't seem very efficient.

Do you have any idea, how this could be accomplished?  
One possibility would be, to change `PositionState`, to also include such an EventEmitter, what do you think?